torchtune/datasets/_chat.py

```diff
@@ -0,0 +1,181 @@
+            formatted_prompt = self.template.format(prompt)
+            encoded_prompt_with_response, labels = tokenize_prompt_and_response(
+                tokenizer=self._tokenizer,
+                prompt=formatted_prompt,
+                response=label,
+                train_on_input=self.train_on_input,
+            )
+            prompt_tokens.extend(encoded_prompt_with_response)
+            label_tokens.extend(labels)
+
+            if len(prompt_tokens) >= self.max_seq_len:
+                break
+
+        prompt_tokens, label_tokens = truncate(
+            self._tokenizer, prompt_tokens, label_tokens, self.max_seq_len
+        )
+
+        assert len(prompt_tokens) == len(label_tokens)
+
+        return prompt_tokens, label_tokens
+
+    def _get_turns(
+        self, dialogue: Dialogue
+    ) -> Generator[Tuple[Dict[str, str], str], None, None]:
+        prompt_messages = {}
+        for message in dialogue:
+            # If we are at the assistant message, we are at the end of a turn, yield.
+            if message["role"] == "assistant":
+                if "user" not in prompt_messages:
+                    raise ValueError(
+                        f"Missing a user message before assistant message: {message['content']}"
+                    )
+                yield prompt_messages, message["content"]
+                prompt_messages = {}
+            # Otherwise, continue to add to the turn's messages
+            else:
+                if message["role"] in prompt_messages:
+                    raise ValueError(
+                        f"Duplicate {message['role']} message in dialogue: {message['content']}"
+                    )
+                prompt_messages[message["role"]] = message["content"]
+
+        # If we never yielded, then the last turn was incomplete
+        if prompt_messages:
+            raise ValueError(
+                f"Incomplete turn in dialogue, current turn: {prompt_messages}"
+            )
+
+
+def chat_dataset(
+    tokenizer: Tokenizer,
+    source: str,
+    conversation_format: str,
+    template: str,
+    max_seq_len: int,
+    train_on_input: bool = False,
+    **load_dataset_kwargs: Dict[str, Any],
+) -> ChatDataset:
+    """
+    Build a configurable dataset with conversations. This method should be
+    used to configure a custom chat dataset from the yaml config instead of
+    using `ChatDataset` directly, as it is made to be config friendly.
+
+    Args:
+        tokenizer (Tokenizer): Tokenizer used to encode data. Tokenize must implement an `encode` and `decode` method.
+        source (str): path string of dataset, anything supported by HuggingFace's `load_dataset`
+            (https://huggingface.co/docs/datasets/en/package_reference/loading_methods#datasets.load_dataset.path)
+        conversation_format (str): string specifying expected format of conversations in the dataset
+            for automatic conversion to the llama format. Supported formats are: "sharegpt"
```

Comment:
Format might be overloaded - here I mean what fields/roles to expect in the data. I purposely enforce users to convert to what our abstractions expect (calling it llama format) so that the subsequent steps can operate on these assumptions (templating, tokenizing, etc assume system role is called "system", user role is called "user" and not "human", etc). This is separate from the prompt template that a model expects, for example using the [INST] tags